File: app/src/main/java/xyz/b515/schedule/util/CourseParser.java
package xyz.b515.schedule.util;

import com.jrummyapps.android.colorpicker.ColorPickerDialog;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.nodes.TextNode;
import org.jsoup.select.Elements;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;
import java.util.Random;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import xyz.b515.schedule.db.CourseManager;
import xyz.b515.schedule.entity.Course;
import xyz.b515.schedule.entity.Spacetime;

public class CourseParser {
    private static final Pattern weekdayTimePattern = Pattern.compile("周(.)第(.+)节\\{第(\\d+)-(\\d+)周\\}");

    public static void parse(String text, CourseManager manager) {
        ArrayList<String> names = new ArrayList<>();

        Document document = Jsoup.parse(text);
        Elements courses = document.getElementById("Table1").getElementsByTag("td");
        for (Element td : courses) {
            if (!td.text().contains("周"))
                continue;
            List<TextNode> nodes = td.textNodes();

            String name = nodes.get(0).text();
            Course course;
            if (!names.contains(name)) {
                names.add(name);

                course = new Course();
                course.setName(name);
                course.setTeacher(nodes.get(2).text());

                int color = ColorPickerDialog.MATERIAL_COLORS[new Random().nextInt(18)];
                course.setColor(ColorHelper.shadeColor(color, 0.33));

                manager.insertCourse(course);
            } else {
                course = manager.getCourse(name);
            }

            Spacetime spacetime = new Spacetime();
            spacetime.setCourse(course);

            Matcher m = weekdayTimePattern.matcher(nodes.get(1).text());
            if (m.find()) {
                spacetime.setWeekday(translateWeekday(m.group(1)));
                String[] weeks = m.group(2).split(",");
                spacetime.setStartTime(Integer.parseInt(weeks[0]));
                spacetime.setEndTime(Integer.parseInt(weeks[weeks.length - 1]));
                spacetime.setStartWeek(Integer.parseInt(m.group(3)));
                spacetime.setEndWeek(Integer.parseInt(m.group(4)));
            }
            //TODO 数据通信原理 {第1-16周|2节/周}

            spacetime.setLocation(nodes.get(3).text());
            if (text.contains("单周")) {
                spacetime.setOddWeek(true);
            } else if (text.contains("双周")) {
                spacetime.setEvenWeek(true);
            }

            manager.insertSpacetime(spacetime);
        }
    }

    private static int translateWeekday(String s) {
        switch (s) {
            case "一":
                return Calendar.MONDAY;
            case "二":
                return Calendar.TUESDAY;
            case "三":
                return Calendar.WEDNESDAY;
            case "四":
                return Calendar.THURSDAY;
            case "五":
                return Calendar.FRIDAY;
            case "六":
                return Calendar.SATURDAY;
            case "日":
                return Calendar.SUNDAY;
        }
        return 0;
    }
}

File: app/src/main/kotlin/xyz/b515/schedule/util/CourseParser.kt
package xyz.b515.schedule.util

import com.jrummyapps.android.colorpicker.ColorPickerDialog
import org.jsoup.Jsoup
import xyz.b515.schedule.db.CourseManager
import xyz.b515.schedule.entity.Course
import xyz.b515.schedule.entity.Spacetime
import java.util.*

object CourseParser {
    private val weekdayTimePattern = "周(.)第(.+)节\\{第(\\d+)-(\\d+)周\\}".toRegex()

    fun parse(text: String, manager: CourseManager) {
        val names = ArrayList<String>()

        val document = Jsoup.parse(text)
        val courses = document.getElementById("Table1").getElementsByTag("td")
        for (td in courses) {
            if ("周" !in td.text())
                continue

            for (nodes in td.textNodes().map { it.text() }.chunk(4)) {
                val name = nodes[0]
                val course: Course
                if (name !in names) {
                    names.add(name)

                    course = Course()
                    course.name = name
                    course.teacher = nodes[2]

                    val color = ColorPickerDialog.MATERIAL_COLORS[Random().nextInt(18)]
                    course.color = ColorHelper.shadeColor(color, 0.33)

                    manager.insertCourse(course)
                } else {
                    course = manager.getCourse(name)
                }

                val spacetime = Spacetime()
                spacetime.course = course

                val m = weekdayTimePattern.find(nodes[1])
                m?.let {
                    val values = m.groupValues
                    val weeks = values[2].split(",")
                    with(spacetime) {
                        weekday = translateWeekday(values[1])
                        startTime = weeks.first().toInt()
                        endTime = weeks.last().toInt()
                        startWeek = values[3].toInt()
                        endWeek = values[4].toInt()
                        location = nodes[3]
                        oddWeek = "单周" in nodes[1]
                        evenWeek = "双周" in nodes[1]
                    }
                }

                manager.insertSpacetime(spacetime)
            }
        }
    }

    private fun translateWeekday(s: String) = when (s) {
        "一" -> Calendar.MONDAY
        "二" -> Calendar.TUESDAY
        "三" -> Calendar.WEDNESDAY
        "四" -> Calendar.THURSDAY
        "五" -> Calendar.FRIDAY
        "六" -> Calendar.SATURDAY
        "日" -> Calendar.SUNDAY
        else -> 0
    }

    fun <T> List<T>.chunk(size: Int): List<List<T>> {
        tailrec fun <T> List<T>._chunk(size: Int, initial: List<List<T>> = emptyList()): List<List<T>> {
            val result = initial + listOf(this.take(size))
            return if (this.size > size) this.drop(size)._chunk(size, result) else result
        }
        return this._chunk(size)
    }
}
